### ISSUE TITLE
YOLOv11-POSE: Improving the ability to locate the coordinates of the keypoints

### ISSUE THREAD

**Issue by harrymai95** [ultralytics/ultralytics#21559]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Could I ask? How to increase the ability to locate the coordinates of the keypoints in the best way? I am also training a similar case, how much should the pose index be increased? Is 100 enough or how much is maximum? Is the kobj index reduced to 0.25 too much?
This is my setting, I don't know if I did anything wrong, hope you can help.
```python
model = YOLO(model_path)
model_trained = model.train(
        #Fix index
        data=data,
        name=name_train,
        project=project_train,
        val=True,
        save=True,
        exist_ok=True,
        plots=True,
        device= 0,
        seed = 42,
        pretrained=True,
        multi_scale = True, 
        #Flexible index
        lr0= 0.001,
        batch= -1,      
        epochs= 200,
        patience= 30,   
        imgsz= 640,
        freeze= 0,
        workers = 8,    
        cache = 'ram',  
        # Augmentation
        hsv_v=0.5,
        degrees = 10.0,
        mosaic=0.5,
        copy_paste=0.5,
        # keypoint-model
        pose = 100.0,       
        kobj = 0.25,        
        box = 0.01, cls = 0.01   
    ) 
``` 
and I have a result as image. Could you give me more advice about how to increase the performance of the navigation keypoint of the model? and Is there anything wrong with my code and result?

<img width="4200" height="1200" alt="Image" src="https://github.com/user-attachments/assets/02c10840-f866-4f39-adae-b54c625b755f" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @harrymai95, thank you for reaching out and sharing your use case with Ultralytics 🚀! This is an automated response to help you get started—an Ultralytics engineer will review your issue soon and provide more specific assistance.

For advanced questions about training and optimizing keypoint detection with YOLO11-POSE, we recommend exploring the [Docs](https://docs.ultralytics.com/) for details on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, as well as tips on hyperparameter tuning and model performance.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug more efficiently.

Since your question relates to custom training, please try to include as much detail as possible. This might include sample images from your dataset, training logs, and confirmation that you are following the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for more support and discussion:
- For real-time help, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth Q&A, try [Discourse](https://community.ultralytics.com/)
- For community-driven threads, check out our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). Upgrading can often resolve unexpected issues:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in several up-to-date, verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience and for contributing to Ultralytics!

**Comment by Y-T-G** [comment]
The loss weights should be balanced. The default values are the best. If you want to increase any of them, you should only do so by a small amount like 1-5.

**Comment by harrymai95** [comment]
> The loss weights should be balanced. The default values are the best. If you want to increase any of them, you should only do so by a small amount like 1-5.

Thanks, I tried it before but it still does not work well, so that why I decrease the loss of class and box low because I just want to focus on keypoint. I saw that others discussed increasing the pose to 40.0 (default is 12.0) and decreasing the kobj to 0.5 (default: 2.0). So I just wonder what is the available range and suitable range of them.

**Comment by glenn-jocher** [comment]
For pose keypoint optimization, extreme values like `pose=100.0` and `kobj=0.25` are counterproductive and will likely hurt overall model performance. The values you mentioned from other discussions (`pose=40.0` and `kobj=0.5`) are much more reasonable.

I'd recommend staying within these ranges:
- `pose`: 12.0-40.0 (default 12.0)
- `kobj`: 0.5-2.0 (default 2.0) 
- Keep `box` and `cls` at or near defaults (7.5 and 0.5 respectively)

Rather than drastically altering loss weights, focus on data quality - ensure your keypoint annotations are precise, increase dataset diversity, and consider using a larger image size like `imgsz=1280` for better keypoint resolution. The model needs balanced learning across all tasks to achieve optimal keypoint performance.

**Comment by harrymai95** [comment]
Hi, sorry to bother you again @glenn-jocher . I’m evaluating a pose model and see metrics like PCK, MPJPE, and OKS mentioned in other topics. Which of these is best to use, and does YOLO support any of them? When I run model.val(), it seems to focus more on bbox than keypoints. Also, how can I calculate these metrics (especially OKS), and display them (as numbers or graphs)? OKS seems most common, but I don’t think it’s included in YOLO.

**Comment by glenn-jocher** [comment]
YOLO does support OKS evaluation! The [kpt_iou function](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.kpt_iou) calculates Object Keypoint Similarity, and the [PoseMetrics class](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.PoseMetrics) handles pose-specific evaluation alongside detection metrics.

When you run `model.val()`, you're getting both box and pose metrics - look for the "Pose(P, R, mAP50, mAP50-95)" columns in your validation output. To access detailed pose metrics programmatically:

```python
results = model.val()
pose_metrics = results.pose  # Access pose-specific metrics
print(results.summary())     # Shows per-class pose metrics including mAP
```

OKS is the standard metric for pose evaluation and is what YOLO uses internally. For pose tasks, focus on the pose mAP values rather than just detection metrics, as these directly measure keypoint localization accuracy using OKS calculations.
